### ISSUE TITLE
Generic response type cannot be used in conjunction with @ApiResponse swagger annotations

### ISSUE THREAD

**Issue by kenotr0n** [micronaut-projects/micronaut-openapi#356]
tl;dr: controllers with generic response types cannot be clarified, described, or extended

I have a controller class with a method similar to this:

```java
    @Get("/getWrapped/{id}")
// --->    @ApiResponse(responseCode="404", content=@Content(@Schema(implementation=String.class)))
    public Wrapped<ResponseType> fetchWrapped(@PathVariable("id") String id) {
        ...
        return Wrapped.of(response);
    }
```

which will produce the 'correct' swagger yaml for the response _so long as there are no @ApiResponse annotations on the method_.  If any swagger annotations for other response codes (404, 500, etc) are included for the controller method, the resulting yaml will not include a response for the 'default' return type (`Wrapped<ResponseType>` in the above case).

Further, if a user wishes to add detail to the default response annotation (other than changing the name of the type), such as a return code of 201, or a description of the nested generic types, this is not possible.

```java
    @ApiResponse(responseCode="201")
    public Wrapped<ResponseType> fetchWrapped(@PathVariable("id") String id) { ... }

```

produces:

```yaml
      responses:
        "201": {}
```

If one wishes to add extra information about a generic response type, the output is a plain `object`:

```java
    @ApiResponse(responseCode="201",content=@Content(schema=@Schema(description="A wrapped type")))
    public Wrapped<ResponseType> fetchWrapped(@PathVariable("id") String id) {
```

produces:

```yaml
        "201": 
          content:
            application/json:
              schema:
                type: object
                description: A wrapped type
```

Attempting to include a generic `implementation` argument to the `@Schema` annotation will only lead to pain and failed compilation; the argument must be a class literal, and generic types cannot be made into class literals.  This fails to compile:

```java
    @ApiResponse(responseCode="201",content=@Content(
            schema=@Schema(implementation=Wrapped<ResponseType>.class)))
```

or any variants attempted thereof.  It looks like `micronaut-openapi` accomplishes this with introspection rather than with the swagger annotations on the controllers.  So if one wishes to expand on the micronaut-generated swagger documentation using _swagger annotations_, the response type must not be generic.

A poor workaround is to copy the response classes into a non-generic concrete type (including all fields, getters, setters, etc).  This works, but violates DRY principles and clutters the namespace with useless classes, adding complexity and maintenance troubles down the road.

I don't foresee the java specification changing to allow generic class literals for source-processing annotation processors.  Perhaps a later version of swagger annotations will include a `nestedSchema` of the inner generic types, but barring these, the closest answer is in `micronaut-openapi` making it possible to work around the deficiencies of official swagger annotations.

### Steps to Reproduce

1. compile example controller including `micronaut-openapi` annotation processor dependency
2. view resulting yaml under `build/classes/java/main/META-INF/swagger`
3. examine response types for example controller's operation

### Expected Behaviour

Operation schema should have a response code 201 with schema type`$ref: '#/components/schema/Wrapped_ResponseType_'` and full description and/or examples.

### Actual Behaviour

Operation swagger yaml cannot include response code, examples, or description of the nested generic type.

### Environment Information

- **Operating System**: MacOS 10.15.7
- **Micronaut Version:** 2.1.0
- **JDK Version:** 11.0.7 AdoptOpenJDK

**Comment by kenotr0n** [comment]
I should add this is a known issue with `swagger-core` as well, it's been reported for years with no suitable resolution.  With over 500 issues on swagger-core, I don't see it being solved there any time soon.  An additional `nestedSchema` method of the `Schema` annotation feels like the right solution to me, akin to the `ArraySchema` method `arraySchema` to mark the properties of the outer array type and `schema` to mark the properties of the inner type.  That's pretty far out of the scope of this project, however.

A general workaround could be implemented in this project around here, where the default response is created:  https://github.com/micronaut-projects/micronaut-openapi/blob/6c60c65c98a01d65f14f5d16f03684d10d461a10/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiEndpointVisitor.java#L632

Perhaps if the `responseCode` is set to `default`, the default sensed schema implementation can be used instead of nothing, so that the generic types are properly represented in the output yaml.  That wouldn't help the case where other `@ApiResponse` annotations are present, or if another `@Schema` annotation needed to refer to a generic wrapper type.  My current sub-optimal workaround is just to leave out all `@ApiResponse` annotations on the methods that return generic types, but this means I can't document the 400- and 500-level errors at the same time.  Anyone have other ideas?

**Comment by urbansky** [comment]
We have the same problem. We comment out all `@ApiResonse`, so we have at least the 200 response with the correct schema.

**Comment by altro3** [comment]
@urbansky @kenotr0n Hi! Since swagger 2.2.0 in annotation `@ApiResponse` you can set `useReturnTypeSchema=true` and write annotations like this:
```java
    @Get("/getWrapped/{id}")
    @ApiResponse(responseCode="404", content=@Content(@Schema(implementation=String.class)))
    @ApiResponse(responseCode="201", useReturnTypeSchema=true)
    public Wrapped<ResponseType> fetchWrapped(@PathVariable("id") String id) {
        return Wrapped.of(response);
    }
```
